### PR TITLE
Add a new output option, workerPublicPath

### PR DESCRIPTION
**Pull request by thomastay**
Presently, Webpack assumes that the worker.js script is located in the same location as the publicPath. The application I work on does not work like that, the worker script and the rest of the script files are located at two separate domains in order to satisfy the Same Origin worker requirement.

Webpack has a feature for making the publicPath a function, unfortunately it is a fairly obscure feature and I can't track down where it applies in the codebase. AFAIK it only applies in AssetGenerator.js to generate the hashes.

There was some work in #14680 to support Cross origin worker loading, however this is unrelated to this change. This PR will not solve cross origin workers. In fact the worker public path option has to be located at the same origin. In our app, the publicPath is a CDN domain, and the worker public path is the same-origin domain.

**What kind of change does this PR introduce?**

Adds in a new output option for the Worker script that tells Webpack where to expect the worker script to be.

**Did you add tests for your changes?**
Done

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**
New option. Documentation is already in the comments.

### CODE REVIEW

**Comment by TheLarkInn** [comment]
We should have a config test for this and add a snapshot to capture the value.

**Review by TheLarkInn — CHANGES_REQUESTED** [review]
LGTM except we should add a case in https://github.com/webpack/webpack/tree/main/test/configCases/output to cover this and then cover it in ConfigCases

File: lib/dependencies/WorkerDependency.js [hunk 1]
```diff
@@ -25,10 +25,13 @@ class WorkerDependency extends ModuleDependency {
...
 	 * @param {string} request request
 	 * @param {[number, number]} range range
+	 * @param {Object} workerDependencyOptions options
+	 * @param {string} workerDependencyOptions.publicPath public path for the worker
 	 */
-	constructor(request, range) {
+	constructor(request, range, workerDependencyOptions) {
 		super(request);
 		this.range = range;
+		this.options = workerDependencyOptions;
```
Comment (sokra) [inline lib/dependencies/WorkerDependency.js:34]: This need to be added to hashing and serialization. Check other dependencies with extra properties for examples

**Review by dhruvilsoni28 — APPROVED** [review]

File: test/__snapshots__/StatsTestCases.basictest.js.snap [hunk 2]
```diff
@@ -4675,3 +4675,25 @@ cacheable modules 2.31 KiB (javascript) 1.37 KiB (webassembly)
     ./node_modules/env.js 34 bytes [built] [code generated]
 webpack x.x.x compiled successfully in X ms"
 `;
+
+exports[`StatsTestCases should print correct stats for worker-public-path 1`] = `
+"With worker public path:
+  asset main-db80575c247d94eb3f64.js 3.53 KiB [emitted] [immutable] (name: main)
```
Comment (thomastay) [inline test/__snapshots__/StatsTestCases.basictest.js.snap:4681]: The test is a bit subtle but as you can see, the main.js file has a different hash from the other main.js file

**Comment by thomastay** [comment]
PTAL, added in a config test as well as a stats test

File: lib/dependencies/WorkerDependency.js (hunk 1 above)
Comment (thomastay) [reply lib/dependencies/WorkerDependency.js:35]: added

**Comment by thomastay** [comment]
p.s. did not mean to remove the review from @alexander-akait , not sure why it removed it

**Comment by TheLarkInn** [comment]
Added back @alexander-akait to reviewers.

**Comment by thomastay** [comment]
Apologies for the lack of recent updates. There is something with the changes that breaks some tests. I was looking into that last week but got sidetracked by some critical tasks that are still ongoing. Intend to get back and fix those tests by this week.

**Comment by thomastay** [comment]
@TheLarkInn PTAL, build is green
@sokra too, addressed your comment

**Review by alexander-akait — APPROVED** [review]

**Review by TheLarkInn — APPROVED** [review]

File: schemas/WebpackOptions.json [hunk 3]
```diff
@@ -5271,6 +5277,10 @@
         }
       },
       "required": ["apply"]
+    },
+    "WorkerPublicPath": {
+      "description": "Worker public path. Much like the public path, this sets the location that the worker script file is intended to be found at. If not set, Webpack will use the publicPath. Don't set this option unless your worker scripts are located at a different path from your other script files.",
```
Comment (snitin315) [inline schemas/WebpackOptions.json:5282]:
```suggestion
      "description": "Worker public path. Much like the public path, this sets the location where the worker script file is intended to be found. If not set, webpack will use the publicPath. Don't set this option unless your worker scripts are located at a different path from your other script files.",
```

**Review by devvsakib — APPROVED** [review]

**Review by devvsakib — APPROVED** [review]

**Comment by TheLarkInn** [comment]
Committing edits and then we will proceed to merge this after CI succeeds.

**Comment by snitin315** [comment]
I think we need to run `yarn special-lint-fix` here

https://github.com/webpack/webpack/blob/4ba225225b1348c8776ca5b5fe53468519413bc0/package.json#L158

**Comment by TheLarkInn** [comment]
> I think we need to run `yarn special-lint-fix` here
> 
> https://github.com/webpack/webpack/blob/4ba225225b1348c8776ca5b5fe53468519413bc0/package.json#L158

@thomastay will you do the following to fix tests.

**Comment by alexander-akait** [comment]
@thomastay sorry for ping, can you do it, because we are planning to do release soon, so you will see this improvement soon :smile:

**Comment by thomastay** [comment]
Hi, I was on vacation the past 2 weeks, will look into this this week

**Comment by alexander-akait** [comment]
@thomastay Sorry for pinging you, looks like we have an old jest bug with newlines and spaces (yeah, we are planning to update jest in near future), can you fix it

**Review by TheLarkInn — APPROVED** [review]
LGTM 🚀 Thank you for the contribution.

**Review by prov1der96 — COMMENTED** [review]
[README.md](https://github.com/webpack/webpack/files/11132281/README.md)

**Comment by prov1der96** [comment]
[README.md](https://github.com/webpack/webpack/files/11132287/README.md)

**Comment by prov1der96** [comment]
[README.md](https://github.com/webpack/webpack/files/11132288/README.md)

**Pull request merged**